### PR TITLE
[PKX-011] Filter Pokemon functionality

### PR DESCRIPTION
**Pull request by giantiyan**

### CODE REVIEW

File: lib/features/home_page/home_page.dart [hunk 1]
```diff
@@ -65,14 +75,48 @@ class HomePage extends StatelessWidget {
...
-        },
-        backgroundColor: red,
-        child: const Icon(Icons.filter_list),
+      floatingActionButton: Row(
+        mainAxisAlignment: MainAxisAlignment.end,
+        children: [
+          Visibility(
+            visible: isVisible ?? false,
+            child: FloatingActionButton(
+              heroTag: null,
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: add unique hero tag like 'cancel' or something

File: lib/features/home_page/home_page.dart [hunk 2]
```diff
@@ -65,14 +75,48 @@ class HomePage extends StatelessWidget {
...
-      floatingActionButton: FloatingActionButton(
-        onPressed: () {
-          // TODO:
-        },
-        backgroundColor: red,
-        child: const Icon(Icons.filter_list),
+      floatingActionButton: Row(
+        mainAxisAlignment: MainAxisAlignment.end,
+        children: [
+          Visibility(
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: You can just use the checking `(isVisible)` to show the FAB. 

`if (isVisible) FloatingActionButton`

File: lib/features/home_page/home_page.dart [hunk 3]
```diff
@@ -65,14 +75,48 @@ class HomePage extends StatelessWidget {
...
+            visible: isVisible ?? false,
+            child: FloatingActionButton(
+              heroTag: null,
+              onPressed: () {
+                onCancel();
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: `onPressed: onCancel`

File: lib/features/home_page/home_page.dart [hunk 4]
```diff
@@ -65,14 +75,48 @@ class HomePage extends StatelessWidget {
...
+              },
+              backgroundColor: red,
+              child: const Icon(Icons.cancel),
+            ),
+          ),
+          SizedBox(width: 8),
+          FloatingActionButton(
+            heroTag: null,
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: add unique hero tag like 'filter' or something

File: lib/features/home_page/home_page.dart [hunk 5]
```diff
@@ -65,14 +75,48 @@ class HomePage extends StatelessWidget {
...
+            onPressed: () {
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: convert to expression body

File: lib/features/home_page/home_page.dart [hunk 6]
```diff
@@ -65,14 +75,48 @@ class HomePage extends StatelessWidget {
...
+              showModalBottomSheet(
+                backgroundColor: Colors.transparent,
+                context: context,
+                isScrollControlled: true,
+                builder: (context) {
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: `builder: (context) => FilterModal(onFilter: onFilter)` should be enough

File: lib/features/home_page/widgets/filter_modal.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  @override
+  Widget build(BuildContext context) {
+
+    return Container(
+        child: Container(
+            padding: EdgeInsets.all(12),
+            decoration: BoxDecoration(
+              color: Colors.white,
+              borderRadius: BorderRadius.vertical(
+                top: Radius.circular(25),
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/filter_modal.dart]: remove trailing comma

File: lib/features/home_page/widgets/filter_modal.dart [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+              ),
+            ),
+            child: Wrap(
+              alignment: WrapAlignment.center,
+              children: <Widget>[
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/filter_modal.dart]: You can remove `<Widget>` here

File: lib/features/home_page/widgets/filter_modal.dart [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                        onFilter(type);
+                        Navigator.pop(context);
+                      },
+                      child: Type(
+                        typeName: type,
+                        pokemonType: type,
+                      ),
+                    ))
+              ],
+            )));
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/filter_modal.dart]: trailing commas here

File: lib/features/home_page/widgets/type.dart [hunk 10]
```diff
@@ -16,23 +16,25 @@ class Type extends StatelessWidget {
     return Column(
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/type.dart]: `Column` widget not needed

File: lib/features/home_page/widgets/type.dart [hunk 11]
```diff
@@ -16,23 +16,25 @@ class Type extends StatelessWidget {
...
             borderRadius: BorderRadius.circular(25),
           ),
-          child: Text(
-            typeName?.toTitleCase ?? '',
-            style: TextStyle(color: white),
+          margin: EdgeInsets.only(
+            bottom: 8,
+            right: 4,
+          ),
+          child: Container(
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/type.dart]: You can use the `Padding` widget here

File: lib/state/actions/home_page_actions.dart [hunk 12]
```diff
@@ -13,6 +14,26 @@ class GetPokemonAction extends LoadingAction {
   @override
   Future<AppState> reduce() async {
     final pokemon = await PokemonHandler.getPokemon();
-    return state.copyWith(pokemon: pokemon);
+    return state.copyWith(
+      pokemon: pokemon,
+      isVisible: false,
```
Comment (ryle-roasa) [inline lib/state/actions/home_page_actions.dart]: Create separate action for updating the `isVisible` property in state. Then just call it in `before() `or `after() `method of this action

File: lib/state/actions/home_page_actions.dart [hunk 13]
```diff
@@ -13,6 +14,26 @@ class GetPokemonAction extends LoadingAction {
...
+      : super(actionKey: filterPokemonListKey);
+
+  final String? pokemonType;
+
+  @override
+  Future<AppState> reduce() async {
+    final pokemon = await PokemonHandler.filterPokemon(pokemonType!);
+    return state.copyWith(
+      pokemon: pokemon,
+      isVisible: true,
```
Comment (ryle-roasa) [inline lib/state/actions/home_page_actions.dart]: Same here

File: lib/utilities/constants.dart [hunk 14]
```diff
@@ -24,3 +26,24 @@ const spinKitDualRing = SpinKitDualRing(
   color: red,
   size: 50.0,
 );
+
+var typesList = [
```
Comment (ryle-roasa) [inline lib/utilities/constants.dart]: `const typeList = [...]`

File: lib/state/actions/home_page_actions.dart [hunk 15]
```diff
@@ -13,6 +14,26 @@ class GetPokemonAction extends LoadingAction {
...
     final pokemon = await PokemonHandler.getPokemon();
-    return state.copyWith(pokemon: pokemon);
+    return state.copyWith(
+      pokemon: pokemon,
+      isVisible: false,
+    );
+  }
+}
+
+/// Get pokemon and save it to the state
```
Comment (ryle-roasa) [inline lib/state/actions/home_page_actions.dart]: Update documentation

File: lib/state/actions/home_page_actions.dart (hunk 12 above)
Comment (ryle-roasa) [reply lib/state/actions/home_page_actions.dart]:
```
@override
  void after() {
    dispatch(ChangeFilterCancelVisibilityAction(isShown: false));
    super.after();
  }
```

File: lib/features/home_page/widgets/filter_modal.dart [hunk 16]
```diff
@@ -9,30 +9,31 @@ class FilterModal extends StatelessWidget {
...
-
     return Container(
-        child: Container(
-            padding: EdgeInsets.all(12),
-            decoration: BoxDecoration(
-              color: Colors.white,
-              borderRadius: BorderRadius.vertical(
-                top: Radius.circular(25),
+      child: Container(
+        padding: EdgeInsets.all(12),
```
Comment (SaturnoMark) [inline lib/features/home_page/widgets/filter_modal.dart]: const

**Review by SaturnoMark — APPROVED** [review]

**Pull request merged**